source/server/admin/stats_handler.cc

```diff
@@ -225,29 +254,82 @@ StatsHandler::statsAsJson(const std::map<std::string, uint64_t>& all_stats,
     stats_array.push_back(ValueUtil::structValue(stat_obj));
   }
 
-  ProtobufWkt::Struct histograms_obj;
-  auto* histograms_obj_fields = histograms_obj.mutable_fields();
-
   ProtobufWkt::Struct histograms_obj_container;
   auto* histograms_obj_container_fields = histograms_obj_container.mutable_fields();
-  std::vector<ProtobufWkt::Value> computed_quantile_array;
 
-  bool found_used_histogram = false;
+  // Display bucket data if histogram_buckets query parameter is used, otherwise output contains
+  // quantile summary data.
+  switch (histogram_buckets_value) {
+  case Utility::HistogramBucketsValue::Null:
+    statsAsJsonQuantileSummaryHelper(*histograms_obj_container_fields, all_histograms, used_only,
+                                     regex);
+    break;
+  case Utility::HistogramBucketsValue::Cumulative:
+    statsAsJsonHistogramBucketsHelper(
+        *histograms_obj_container_fields, all_histograms, used_only, regex,
+        [](const Stats::ParentHistogramSharedPtr& histogram) {
+          return histogram->intervalStatistics().computedBuckets();
+        },
+        [](const Stats::ParentHistogramSharedPtr& histogram) {
+          return histogram->cumulativeStatistics().computedBuckets();
+        });
+    break;
+  case Utility::HistogramBucketsValue::Disjoint:
+    statsAsJsonHistogramBucketsHelper(
+        *histograms_obj_container_fields, all_histograms, used_only, regex,
+        [](const Stats::ParentHistogramSharedPtr& histogram) {
+          return histogram->intervalStatistics().computeDisjointBuckets();
+        },
+        [](const Stats::ParentHistogramSharedPtr& histogram) {
+          return histogram->cumulativeStatistics().computeDisjointBuckets();
+        });
+    break;
+  }
+
+  // Add histograms to output if used histogram found.
+  if (histograms_obj_container_fields->contains("histograms")) {
+    stats_array.push_back(ValueUtil::structValue(histograms_obj_container));
+  }
+
+  auto* document_fields = document.mutable_fields();
+  (*document_fields)["stats"] = ValueUtil::listValue(stats_array);
+
+  return MessageUtil::getJsonStringFromMessageOrDie(document, pretty_print, true);
+}
+
+std::string
+StatsHandler::computeDisjointBucketSummary(const Stats::ParentHistogramSharedPtr& histogram) {
+  if (!histogram->used()) {
+    return "No recorded values";
+  }
+  std::vector<std::string> bucket_summary;
+  Stats::ConstSupportedBuckets& supported_buckets =
+      histogram->intervalStatistics().supportedBuckets();
+  const std::vector<uint64_t> disjoint_interval_buckets =
+      histogram->intervalStatistics().computeDisjointBuckets();
```

Comment:
put `histogram->intervalStatistics()` in a temp since you reference it twice.

Reply:
Done

Reply:
nit: doesn't look done :)

Reply:
My bad :), I think I changed it in all places now